src/_pytest/logging.py

```diff
@@ -527,6 +527,11 @@ def __init__(self, config: Config) -> None:
         # File logging.
         self.log_file_level = get_log_level_for_setting(config, "log_file_level")
         log_file = get_option_ini(config, "log_file") or os.devnull
+
+        # Keep the log file relative to the inidir file (if it exists) (#7336)
+        inidir = getattr(getattr(config, "inifile", None), "dirname", None)
```

Comment:
I realize this is a matter of opinion, but this seems a bit contrived, I would instead use this:

```python
if log_file != os.devnull and config.inifile:
    log_file = os.path.join(config.inifile.dirname, log_file)
```

(No need for the `getattr` chain)

Reply:
Thanks, any advice on fixing:

```
src\_pytest\logging.py:533: error: "str" has no attribute "dirname"  [attr-defined]
Found 1 error in 1 file (checked 2 source files)
``` 

As a result?

Reply:
That's weird, `locate_config` returns:

```python
Tuple[
    Optional[py.path.local], Optional[py.path.local], Dict[str, Union[str, List[str]]],
]
```

(Note that the 2nd item is `Optional[py.path.local]`)

Here's a snippet from `determine_setup`, which calls `locate_config`:

```
def determine_setup(
    inifile: Optional[str],
    args: List[str],
    rootdir_cmd_arg: Optional[str] = None,
    config: Optional["Config"] = None,
) -> Tuple[py.path.local, Optional[str], Dict[str, Union[str, List[str]]]]:
    rootdir = None
    dirs = get_dirs_from_args(args)
    if inifile:  # (1)
        inicfg = load_config_dict_from_file(py.path.local(inifile)) or {}
        if rootdir_cmd_arg is None:
            rootdir = get_common_ancestor(dirs)
    else:  # (2)
        ancestor = get_common_ancestor(dirs)
        rootdir, inifile, inicfg = locate_config([ancestor])
        if rootdir is None and rootdir_cmd_arg is None:
            ...
    return rootdir, inifile, inicfg or {}
```

In `1`, indeed `inifile` ends up as a string by the time of `return`, but not when we get into `2`, where `inifile` is a `py.path.local`.

Shouldn't mypy catch that? 

Regardless, I think we should change the return value from `determine_setup` to always return a `py.path.local`.